### PR TITLE
feat(grpc): add opt-in connection sharing across VUs

### LINKED ISSUE #3188: [grpc] Configurable `grpc.ClientConn` sharing
### Feature Description

Connections in gRPC are able to be shared across multiple executing contexts (whether those contexts are goroutines, threads, etc). This sharing can safely continue to a point where the number of active channels reaches the server-configured maximum number of open streams available to a single http2 connection. Typically, this maximum is set to 100 concurrent streams.

Currently, a K6 gRPC Client is created for every VU. The Client struct has a wrapper for a `grpc.ClientConn`. In a typical test situation, it is common for VUs to communicate with a small number of endpoints. These endpoints are commonly the same endpoint multiplexing gRPC services.

K6 should support the option that would allow test writers to decide if VUs within a test should share gRPC connections to the same endpoint. One of the benefits of connection sharing is that it does not matter if an endpoint is used for two different services. That is not a property of the gRPC connection itself. That is a property of the invocation. So effectively, multiple VUs can use a single gRPC connection so long as that exact endpoint multiplexes the various gRPC services needed by the VUs.

A small change to the core gRPC connection logic can implement this in a safe and concurrent way.

(I've made this change in my personal fork and have tested it in our ecosystem with materially less resource consumption)

One of the key reasons why I, in particular, want this feature is because it substantially reduces the overhead for each VU using a gRPC Client. Instead of opening 50 gRPC Connections using 50 VUs, with this feature in place, those same 50 VUs can safely use the same single open gRPC connection. 

### Suggested Solution (optional)

My solution:

1. Add `ConnectionSharing` to `connectParams`
2. Conditionally call `grpcext.Dial` or `grpcext.DialShared` (new function so the original is untouched)
  a. One way to do this: 
```go
	// Create a local function variable to point to the appropriate Dial
	dialFn := grpcext.Dial
	if p.ConnectionSharing {
		dialFn = grpcext.DialShared
	}

	c.conn, err = dialFn(ctx, addr, opts...)
```
3. Add supporting logic to `grpcext` package with limited requirements:
  a. Must be safe for concurrent access
  b. Create `grpc.ClientConn` once and store the reference for reuse if `ConnectionSharing` is enabled
  c. Clean up reference to `grpc.ClientConn` upon `Close`

Implementation (with tests): https://github.com/chrismoran-mica/k6/tree/feature/cross-vu-shared-grpc-clientconn

### Already existing or connected issues / PRs (optional)

This is my PR for this very feature.

https://github.com/grafana/k6/pull/3179

### PR DESCRIPTION
**Pull request by moko-poi**
## Summary

- Add `connectionSharing` parameter to `grpc.Client.connect()` to allow multiple VUs to share a single `grpc.ClientConn`
- Connection pool is managed by `RootModule` (one per test run) with reference counting for safe cleanup
- gRPC's `ClientConn` handles HTTP/2 stream multiplexing internally, so no per-connection client limit is needed

### Usage

```javascript
import grpc from 'k6/net/grpc';

const client = new grpc.Client();

export default () => {
  client.connect('localhost:50051', {
    plaintext: true,
    connectionSharing: true,
  });

  // All VUs share the same underlying grpc.ClientConn
  const response = client.invoke('my.Service/Method', { key: 'value' });
  client.close();
};
```

### Related

- Closes grafana/k6#3188
- Previous attempts: grafana/k6#3179, grafana/xk6-grpc#31 (both closed)
- xk6-grpc is now archived, so this is implemented directly in k6

## Test plan

- [x] `go test -race ./internal/js/modules/k6/grpc/...` passes
- [x] `golangci-lint run ./internal/js/modules/k6/grpc/...` passes (no new warnings)
- [x] Verify connection sharing reduces TCP connections with a real gRPC server

### CODE REVIEW

File: internal/js/modules/k6/grpc/conn_pool.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+func (p *connectionPool) getOrDial(
+	ctx context.Context,
+	addr string,
+	types *protoregistry.Types,
+	opts ...grpc.DialOption,
+) (*grpcext.Conn, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	key := strings.ToLower(addr)
```
Comment (oleiade) [inline internal/js/modules/k6/grpc/conn_pool.go:46]: My understanding is that grpc connections `ClientConn` are not only bound to `host:port`, but also carry TLS config, auth override, call limits, and other behaviors.

I'm a bit concerned that we key the connections based solely on the address (`host:port`) here, as the first client setting a connection to that given address would also essentially set those additional information (TLS, etc) for all the further uses as well. Concretely, a later client asking for different TLS credentials or a different authority would silently get the first client's connection instead?

I'm not sure how much of a problem that could be, as I'm not very experienced with gRPC myself though. Let me know what you folks think 🙇‍♂️

File: internal/js/modules/k6/grpc/conn_pool.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+// getOrDial returns an existing shared connection for the given address if one
+// exists. Otherwise, it dials a new connection and stores it in the pool.
+func (p *connectionPool) getOrDial(
+	ctx context.Context,
+	addr string,
+	types *protoregistry.Types,
+	opts ...grpc.DialOption,
+) (*grpcext.Conn, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
```
Comment (oleiade) [inline internal/js/modules/k6/grpc/conn_pool.go:44]: Do we need to hold the mutex for the whole duration of the function?

File: internal/js/modules/k6/grpc/conn_pool_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+func TestConnectionPool_Release_NonExistent(t *testing.T) {
+	t.Parallel()
+
+	pool := newConnectionPool()
+	err := pool.release("nonexistent:50051")
+	require.NoError(t, err)
+}
+
+func TestConnectionPool_CaseInsensitiveKey(t *testing.T) {
```
Comment (oleiade) [inline internal/js/modules/k6/grpc/conn_pool_test.go:28]: This test's name hints at testing the pool is case insensitve, yet the content of the test only tests lowercase cases?

File: internal/js/modules/k6/grpc/conn_pool_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+	pool.mu.Lock()
+	sc, ok := pool.conns["localhost:50051"]
+	pool.mu.Unlock()
+
+	assert.True(t, ok)
+	assert.Equal(t, 1, sc.refCount)
+}
+
+func TestConnectionPool_ReleaseDecrementsRefCount(t *testing.T) {
```
Comment (oleiade) [inline internal/js/modules/k6/grpc/conn_pool_test.go:48]: I would find it useful to also test with a refcount of zero.

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 1 above)
Comment (joanlopez) [reply internal/js/modules/k6/grpc/conn_pool.go:46]: I have no much experience with gRPC either, but I guess what you said @oleiade makes sense - i.e., might be safer to add more info to the "connection hash" so connection sharing is more specific.

A reason why I could see this acceptable as-is, is because this will remain as opt-in, so at least the user would need to explicitly set it, and thus unexpected behaviors, if any, hopefully will result less confusing. Perhaps this should be an experimental feature, so we remain flexible to change this behavior? 🤔

**Comment by joanlopez** [comment]
Probably not a blocker, but another thing we should probably clarify, and perhaps explicitly document, is that this won't make possible to share connections across all VUs in distributed executions, which I guess is not a problem considering that one of the main reasons stated in #3188 is to reduce the overhead of having a different connection per VU, but still important to consider.

What do you think @oleiade?

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 1 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:46]: You're both right, this is a real problem as it stands. In `Connect()` all the connection-defining settings go into the dial options: TLS config, plaintext vs. TLS, `authority`, max receive/send sizes, and user-agent. But the pool keys only on the address, so a second client asking for different TLS credentials or a different `:authority` on the same `host:port` would silently reuse the first client's connection. This also lines up with a known gRPC-Go constraint: since these settings are bound to the `ClientConn` at dial time, connections can't safely be shared across clients that differ in them (https://github.com/grpc/grpc-go/issues/195).

I'll fix this by keying the pool on the full set of connection-defining parameters instead of the address alone, so only clients with truly identical connection semantics share a `ClientConn` and any mismatch gets its own connection. On the experimental question, I'm happy to also document this as subject to change so we keep room to adjust the sharing semantics later.

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 2 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:44]: Good question, and it's more significant than it first looks. k6's `grpcext.DefaultOptions` dials with `grpc.WithBlock()`, so `grpcext.Dial` blocks until the handshake completes. Holding the mutex across the whole function therefore serializes every VU's `connect()` behind each other's network handshake, even for different addresses. I'll restructure so the global lock only guards the map lookup/insert, and concurrent dials to different keys can proceed in parallel. To avoid two VUs dialing the same new key at once and leaking a connection, I'll insert a placeholder under the lock and have concurrent callers wait on a shared dialing future rather than the global mutex.

File: internal/js/modules/k6/grpc/conn_pool_test.go (hunk 3 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool_test.go:28]: You're right, the test name overpromises since it only stores and reads back the lowercase key. I'll change it to insert with one casing and look up with a different casing, asserting both resolve to the same entry.

File: internal/js/modules/k6/grpc/conn_pool_test.go (hunk 4 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool_test.go:48]: Agreed. I'll add a test that drives the refcount down to zero and asserts the entry is removed and the connection is closed. The current tests use a nil conn to sidestep `Close()`, so I'll add a small closable fake to cover the zero-refcount path end to end.

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 1 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:46]: Agreed on making the sharing more specific. As noted in the thread above, I'll key the pool on the full set of connection-defining parameters so mismatched TLS/authority/etc. never share a connection, and I'll also document the option as experimental and subject to change so we keep room to adjust the semantics later.

**Comment by joanlopez** [comment]
Hi @moko-poi, as you'll be working on the comments above, I also want to take the opportunity for you to take a look at this: https://community.grafana.com/t/action-required-signed-commits-mandatory-for-all-grafana-repositories/163404.

**TL;DR All commits must have verified signatures,** and I see your first commit in this branch isn't, so can you please also re-write the history for this PR to leave it only with signed commits, please? Otherwise, even if you work on the comments from @oleiade, we won't be able to move forward and merge it anyway, unfortunately.

Apologies for the noise, but I hope it's for good! 👍🏻 Thanks! 🙇🏻

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 1 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:46]: Fixed in https://github.com/grafana/k6/pull/6210. The pool is now keyed on every parameter that gRPC binds to the ClientConn at dial time: the address, plaintext, the tls configuration, the authority and the maximum call sizes. Two clients that disagree on any of them get separate connections, so the first client to reach a server can no longer impose its transport credentials or its authority on the ones that come after it.

While doing this I found a related problem in the same area. The shared connection also carried the protobuf type registry of whichever client dialed first, so the other clients would resolve Any values and extensions through a registry that is not their own. Each client now keeps its own registry over the shared connection.

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 2 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:44]: Fixed in https://github.com/grafana/k6/pull/6210. The pool mutex now only guards the map. A client that finds no entry inserts a placeholder and dials outside the lock, and the clients asking for the same connection in the meantime wait on that entry rather than on the pool, so a slow handshake no longer blocks connect calls to other servers. A dial that fails is removed from the pool, so the next client retries it instead of inheriting the error.

File: internal/js/modules/k6/grpc/conn_pool_test.go (hunk 3 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool_test.go:28]: Fixed in https://github.com/grafana/k6/pull/6210. You were right, the test only stored and read back a lowercase key. It now builds the key from two spellings of the same address and asserts they are equal, and then dials the same server through the pool as localhost and as LOCALHOST and asserts that a single connection is created.

File: internal/js/modules/k6/grpc/conn_pool_test.go (hunk 4 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool_test.go:48]: Fixed in https://github.com/grafana/k6/pull/6210. The pool tests now run against a real gRPC server exposing the health service, so a connection can actually be dialed and closed instead of being faked with a nil value. The new test takes two references, releases both, and asserts that the entry is gone from the pool and that both views of the connection fail a health check afterwards. There are also new tests for concurrent dials to the same address and for a dial that fails.

File: internal/js/modules/k6/grpc/conn_pool.go (hunk 1 above)
Comment (moko-poi) [reply internal/js/modules/k6/grpc/conn_pool.go:46]: Both points are addressed in https://github.com/grafana/k6/pull/6210. The key now covers all the parameters that define a connection rather than the address alone, and the option is documented as experimental, noting that the conditions under which a connection is shared may change.

On your other question about distributed execution, sharing happens within a k6 instance, so each instance keeps its own set of connections. That is now stated in the same comment on the option, and in the pull request description.

**Comment by moko-poi** [comment]
Superseded by https://github.com/grafana/k6/pull/6210.

That pull request carries the same feature on top of the current master as a single signed commit, so the unsigned commit that was blocking this one is gone. It also addresses the review here: the pool is keyed on all the parameters that define a connection instead of the address alone, dialing no longer happens under the pool lock, each client keeps its own protobuf type registry over the shared connection, and the pool tests now run against a real gRPC server. The option is documented as experimental and as being local to a k6 instance.

Thank you both for the review, and sorry for moving the discussion to a new pull request. Closing this one in favour of 6210.
